Mathlib/Analysis/Calculus/Gradient.lean

```diff
@@ -0,0 +1,472 @@
+/-
+Copyright (c) 2023 Ziyu Wang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yu Penghao(PKU), Ziyu Wang(PKU), Chenyi Li(PKU), Cao Zhipeng(HUST)
+-/
+import Mathlib.Analysis.InnerProductSpace.PiL2
+import Mathlib.Analysis.Calculus.MeanValue
+import Mathlib.Analysis.InnerProductSpace.Dual
+
+/-!
+# Gradient
+
+## Main Definitions
+
+Let `f` be a function from `Hilbert Space E` to `ℝ`, `x` be a point in E
+and `gradf` be a vector in E. Then
+
+  `HasGradnAt f gradf x`
+
+says that `f` has gradient `gradf'` at `x`, which means:
+
+  `f y - f x = ⟨ gradf , y - x ⟩ + o (y - x)`
+
+Such defintions are based on the definitions of `fderiv`, and can be transformed to each other
+for functions `f` from `Hilbert Space E` to `ℝ `.
+
+## Main results
+
+This file contains the following parts of gradient.
+* the definition of gradient.
+* the theorems translating between `HasGradnAt` and `HasFDerivAt`
+* equivalent theorems translating the `LittleO` notation in `HasGradnAt` to
+  the language of `ε - δ` convergence, easier to be calculated.
+* theorems of the basic calculations of gradient, which can be found as well for `fderiv`.
+* theorems connecting `HasGradnAt`, `ContinuousAt` and `ε - δ` convergence.
+* theorems showing that for the differiential function,
+  the gradient (if exists) at local minima is zero.
+-/
+
+open InnerProductSpace
+noncomputable section
+
+variable {E : Type*} [NormedAddCommGroup E] [InnerProductSpace ℝ E] [CompleteSpace E]
+variable {f: E → ℝ} {f': E → (E →L[ℝ] ℝ)} {x y : E} {s : Set E}
+variable {x y gradient : E} {gradient' : E →L[ℝ] ℝ}
+local notation "⟪" x ", " y "⟫" => @inner ℝ _ _ x y
+
+/-
+### Definition of gradient
+-/
+/-- Translate `g` in `Hilbert Space E` to its dual ℝ-ContinuousLinearMap : `⟨ g , ⬝ ⟩`-/
+local notation "∇*" gradient  => (toDualMap ℝ _) gradient
+/-- Translate the ℝ-ContinuousLinearMap on `Hilbert Space E` to
+  its dual element on E by Frechet-Riesz Represent Theorem -/
+def grad : (E →L[ℝ] ℝ) → E
+  | gradient => ((toDual ℝ E).symm gradient)
+
+theorem grad_equiv :
+  ∀ x y :E, inner (grad (f' x)) y = f' x y := by
+    intro x y; apply toDual_symm_apply
+
+/-- The gradient of a function `f` from Hilbert Space E to ℝ at a point `x`
+    is defined to be the dual of `fderiv` of `f` at `x` -/
+def gradn (f : E → ℝ) (x : E) : E := grad (fderiv ℝ f x)
```

Comment:
We abbreviate only when the the full-length wording is very long.  Here, it's not so long, so I would write in full: `Gradient`.  Same for `HasGradientAt`.